### PR TITLE
[Dashboard] Agent in minimal ray installation

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

This is the second part of https://docs.google.com/document/d/12qP3x5uaqZSKS-A_kK0ylPOp0E02_l-deAbmm8YtdFw/edit#. After this PR, dashboard agents will fully work with minimal ray installation.

Screen Shot 2022-01-24 at 2 53 41 AM

Note that this PR requires to introduce "aioredis", "frozenlist", and "aiosignal" to the minimal installation. These dependencies are very small (or will be removed soon), and including them to minimal makes thing very easy. Please see the below for the reasoning. 

Screen Shot 2022-01-24 at 2 52 26 AM

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: dashboard/utils.py [hunk 1]
```diff
@@ -61,6 +70,15 @@ async def run(self, server):
         :param server: Asyncio GRPC server.
         """
 
+    @staticmethod
+    @abc.abstractclassmethod
+    def is_optional():
```
Comment (rkooo567) [inline dashboard/utils.py:75]: Prefer to have this static method for better maintainability. Alternative way is to just not load modules if it fails to import, but I think this way is more explicit and easy to figure out what modules are loaded with minimal installation.

File: python/ray/_private/services.py [hunk 2]
```diff
@@ -1299,9 +1299,7 @@ def start_dashboard(require_dashboard,
                     raise e
 
         # Make sure the process can start.
-        try:
-            import ray.dashboard.optional_deps  # noqa: F401
-        except ImportError:
+        if not ray._private.utils.check_dashboard_dependencies_installed():
```
Comment (rkooo567) [inline python/ray/_private/services.py:1302]: Unrelated to this PR, but this is a small fix

File: dashboard/utils.py [hunk 3]
```diff
@@ -44,6 +44,15 @@ async def run(self, server):
...
```
Comment (jjyao) [inline dashboard/utils.py:49]: should we name it `is_minimal` or something similar to be consistent with agent

File: dashboard/utils.py [hunk 4]
```diff
@@ -82,11 +100,24 @@ def get_all_modules(module_type):
...
-    return [
-        m for m in module_type.__subclasses__()
-        if getattr(m, "__ray_dashboard_module_enable__", True)
-    ]
+        try:
+            importlib.import_module(name)
+        except ModuleNotFoundError:
+            pass
+
+    should_load_minimal_modules = (
```
[the remark this answers is not in the document]
Comment (architkulkarni) [reply dashboard/utils.py:108]: Agree with this (flips the meaning of the variable, but it's a bit more readable.) Another option is `should_only_load_minimal_modules`)

**Review by architkulkarni — COMMENTED** [review]
Looks good to me pending @jjyao's comments. 

Other minor comments:
- The purpose of `test_runtime_env_ray_minimal.py` is just to test for the `"ray[default]"` error message, so should we just delete this file entirely in this PR?
- Should we remove `"ray[default]"` from this line in this PR?  It installs the appropriate version of `ray["default"]` into the user-specified conda environment.  https://github.com/ray-project/ray/blob/c288b97e5f6f0d376b074927e3ff5596ff902dcc/python/ray/_private/runtime_env/conda.py#L217

File: .buildkite/pipeline.yml [hunk 5]
```diff
@@ -244,6 +244,12 @@
       python/ray/tests/test_output
     - bazel test --test_output=streamed --config=ci --test_env=RAY_MINIMAL=1 $(./scripts/bazel_export_options)
       python/ray/tests/test_runtime_env_ray_minimal
+    - bazel test --test_output=streamed --config=ci $(./scripts/bazel_export_options)
+      python/ray/tests/test_runtime_env
+    - bazel test --test_output=streamed --config=ci $(./scripts/bazel_export_options)
+      python/ray/tests/test_runtime_env_complicated
+    - bazel test --test_output=streamed --config=ci $(./scripts/bazel_export_options)
+      python/ray/tests/test_runtime_env_validation
```
Comment (architkulkarni) [inline .buildkite/pipeline.yml:252]: Why not move all runtime_env tests here instead of just these three?  They should all pass in this PR right?

Comment (rkooo567) [reply .buildkite/pipeline.yml:252]: Can you help finding a list of them?

File: dashboard/utils.py (hunk 3 above)
Comment (rkooo567) [reply dashboard/utils.py:49]: Is_minimal sgtm$

File: dashboard/utils.py [hunk 6]
```diff
@@ -82,11 +100,24 @@ def get_all_modules(module_type):
...
             ray.dashboard.modules.__name__ + "."):
-        importlib.import_module(name)
-    return [
-        m for m in module_type.__subclasses__()
-        if getattr(m, "__ray_dashboard_module_enable__", True)
-    ]
+        try:
+            importlib.import_module(name)
+        except ModuleNotFoundError:
+            pass
```
[the remark this answers is not in the document]
Comment (rkooo567) [reply dashboard/utils.py:106]: Ah this is legacy (before adding is_optional). I will remove it

File: dashboard/utils.py [hunk 7]
```diff
@@ -82,11 +100,24 @@ def get_all_modules(module_type):
...
+
+    should_load_minimal_modules = (
+        not ray._private.utils.check_dashboard_dependencies_installed())
+    imported_modules = []
+    # module_type.__subclasses__() should contain modules that
+    # we could successfully import.
+    for m in module_type.__subclasses__():
+        if not getattr(m, "__ray_dashboard_module_enable__", True):
```
[the remark this answers is not in the document]
Comment (rkooo567) [reply dashboard/utils.py:114]: It is used to entirely turn on / off “module” in the dashboard. I think it was added as feature flag when the module is introduced first time.

File: .buildkite/pipeline.yml (hunk 5 above)
Comment (architkulkarni) [reply .buildkite/pipeline.yml:252]: Yup, here's a list, all in `/ray/tests`:
```
test_runtime_env.py
test_runtime_env_complicated.py
test_runtime_env_conda_and_pip.py
test_runtime_env_conda_and_pip_2.py
test_runtime_env_conda_and_pip_3.py
test_runtime_env_env_vars.py
test_runtime_env_fork_process.py
test_runtime_env_packaging.py
test_runtime_env_plugin.py
test_runtime_env_ray_minimal.py
test_runtime_env_validation.py
test_runtime_env_working_dir.py
test_runtime_env_working_dir_2.py
test_runtime_env_working_dir_3.py
test_runtime_env_working_dir_remote_uri.py
```

But after looking at it a bit more, I'm not sure if it's the right decision to include them all here.  It looks like the "Minimal install" test is only used as a kind of sanity check (it only runs some test_basic workloads).  If that's the case, then just running the three tests you added should be fine.

Comment (rkooo567) [reply .buildkite/pipeline.yml:252]: I think the only concern we have is if any of those tests require extra dependencies. I believe the existing set of tests should cover this, but you can give me confirmation!

File: dashboard/utils.py (hunk 3 above)
Comment (rkooo567) [reply dashboard/utils.py:49]: Going with `is_minimal_module`

File: dashboard/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rkooo567) [reply dashboard/utils.py:108]: Prefer should_only_load_minimal_modules. It is easier to read if statement underneath with this

```
        if should_only_load_minimal_modules and not m.is_minimal_module():
            continue
```

**Comment by rkooo567** [comment]
Comments addressed

File: dashboard/agent.py [hunk 8]
```diff
@@ -177,44 +176,19 @@ async def _check_parent():
...
+            # to communicate with the dashboard in a head node.
+            # Http server is not started in the minimal version because
+            # it requires additional dependencies that are not
+            # included in the minimal ray package.
+            self.http_server = await self._configure_http_server(modules)
 
         # Write the dashboard agent port to redis.
         # TODO: Use async version if performance is an issue
+        # -1 should indicate that http server is not started.
+        http_port = -1 if not self.http_server else self.http_server.http_port
```
Comment (jjyao) [inline dashboard/agent.py:191]: really nit: None is probably better

**Review by jjyao — APPROVED** [review]

File: .buildkite/pipeline.yml (hunk 5 above)
Comment (architkulkarni) [reply .buildkite/pipeline.yml:252]: Yup, existing tests are fine

**Review by architkulkarni — APPROVED** [review]

**Comment by rkooo567** [comment]
@architkulkarni 

> The purpose of test_runtime_env_ray_minimal.py is just to test for the "ray[default]" error message, so should we just delete this file entirely in this PR?

The test seems okay for very minimal functionality? Up to you if we should delete it or not.

> Should we remove "ray[default]" from this line in this PR? It installs the appropriate version of ray["default"] into the user-specified conda environment.

I will keep it for now (in case this can cause regressions). Please feel free to change this if you think it is necessary!

**Comment by rkooo567** [comment]
@edoakes @richardliaw 

Can you guys give approval on new dependencies? Or should we have some conversation from the api channel?

**Comment by edoakes** [comment]
The delta is `"aioredis < 2", "aiosignal", "frozenlist",`, right? They all look pretty modest, so looks ok to me.

**Comment by rkooo567** [comment]
@edoakes can you give the approval on that? I need one of you guys approval to merge the PR!

File: python/setup.py [hunk 9]
```diff
@@ -258,18 +255,12 @@ def get_packages(self):
...
-        "numpy >= 1.19.3; python_version >= '3.9'",
-        "protobuf >= 3.15.3",
-        "pyyaml",
-        "redis >= 3.5.0",
+        "attrs", "click >= 7.0", "dataclasses; python_version < '3.7'",
+        "filelock", "grpcio >= 1.28.1", "jsonschema",
+        "msgpack >= 1.0.0, < 2.0.0", "numpy >= 1.16; python_version < '3.9'",
+        "numpy >= 1.19.3; python_version >= '3.9'", "protobuf >= 3.15.3",
+        "pyyaml", "redis >= 3.5.0", "aioredis < 2", "aiosignal", "frozenlist",
+        ""
```
Comment (richardliaw) [inline python/setup.py:263]: why is there an empty thing here?

Comment (rkooo567) [reply python/setup.py:263]: Maybe a mistake. Let me remove it

Comment (rkooo567) [reply python/setup.py]: Fixed

Comment (richardliaw) [reply python/setup.py]: Not the biggest fan of the dependency shift but I guess we can move forward with it.

**Review by richardliaw — APPROVED** [review]

**Comment by rkooo567** [comment]
Failed tests seem unrelated

**Pull request merged**